### PR TITLE
[Feature] Support VideoMAE

### PR DESCRIPTION
**Pull request by hukkai**
The test accuracies are matched with `SampleFramesV2`

### CODE REVIEW

**Comment by hukkai** [comment]
@Dai-Wenxun Please review

File: mmaction/models/backbones/vit_mae.py [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+            `dict(type='LN', eps=1e-6)`.
+        init_cfg (dict | None): The Config for initialization. Defaults to
+            None.
+    """
+
+    def __init__(self,
+                 embed_dims: int,
+                 num_heads: int,
+                 mlp_ratio: int = 4.,
+                 qkv_bias: bool = False,
```
Comment (Dai-Wenxun) [inline mmaction/models/backbones/vit_mae.py:127]: Default value does not match docstring.

File: mmaction/models/backbones/vit_mae.py [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+        if hasattr(self, 'gamma_1'):
+            x = x + self.drop_path(self.gamma_1 * self.attn(self.norm1(x)))
+            x = x + self.drop_path(self.gamma_2 * self.mlp(self.norm2(x)))
+        else:
+            x = x + self.drop_path(self.attn(self.norm1(x)))
+            x = x + self.drop_path(self.mlp(self.norm2(x)))
+        return x
+
+
+def get_sinusoid_encoding(n_position: int, embed_dims: int):
```
Comment (Dai-Wenxun) [inline mmaction/models/backbones/vit_mae.py:186]: The type hint for the return value is missing.

File: mmaction/models/backbones/vit_mae.py [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+        num_heads (int): Number of parallel attention heads.
+        qkv_bias (bool): If True, add a learnable bias to q and v.
+            Defaults to True.
+        qk_scale (float | optional): Override default qk scale of
+            ``head_dim ** -0.5`` if set. Defaults to None.
+        attn_drop_rate (float): Dropout ratio of attention weight.
+            Defaults to 0.
+        drop_rate (float | optional): Dropout ratio of output. Defaults to 0.
+        init_cfg (dict | None): The Config for initialization. Defaults to
+            None.
```
Comment (Dai-Wenxun) [inline mmaction/models/backbones/vit_mae.py:33]:
```suggestion
        init_cfg (dict | None): The Config for initialization. 
            Defaults to None.
```

File: mmaction/models/backbones/vit_mae.py [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+        drop_path_rate (float): Dropout ratio of the residual branch.
+            Defaults to 0.
+        init_values (float): Value to init the multiplier of the residual
+            branch. Defaults to 0.
+        act_cfg (dict): Config for activation layer in FFN. Defaults to
+            `dict(type='GELU')`.
+        norm_cfg (dict): Config for norm layers. Defaults to
+            `dict(type='LN', eps=1e-6)`.
+        init_cfg (dict | None): The Config for initialization. Defaults to
+            None.
```
Comment (Dai-Wenxun) [inline mmaction/models/backbones/vit_mae.py:120]:
```suggestion
        init_cfg (Union[dict, :obj:`mmengine.ConfigDict`], optional): The
            config for initialization. Defaults to None.
```

File: mmaction/models/backbones/vit_mae.py [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+        init_values (float): Value to init the multiplier of the residual
+            branch. Defaults to 0.
+        use_learnable_pos_emb (bool): If True, use learnable positional
+            embedding, othersize use sinusoid encoding. Defaults to False.
+        num_frames (int): Number of frames in the video. Defaults to 16.
+        tubelet_size (int): Temporal size of one patch. Defaults to 2.
+        use_mean_pooling (bool): If True, take the mean pooling over all
+            positions. Defaults to True.
+        init_cfg (dict | None): The Config for initialization. Defaults to
+            None.
```
Comment (Dai-Wenxun) [inline mmaction/models/backbones/vit_mae.py:247]:
```suggestion
        init_cfg (Union[dict, :obj:`mmengine.ConfigDict`], optional): The
            config for initialization. Defaults to None.
```

File: mmaction/models/backbones/vit_mae.py [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+            self.gamma_2 = nn.Parameter(
+                init_values * torch.ones((dim)), requires_grad=True)
+
+    def forward(self, x: Tensor) -> Tensor:
+        """Defines the computation performed at every call.
+
+        Args:
+            x (Tensor): The input data with size of (B, N, C).
+        Returns:
+            Tensor: The output of the attention block, same size as inputs.
```
Comment (Dai-Wenxun) [inline mmaction/models/backbones/vit_mae.py:175]: attention -> basic

File: mmaction/models/backbones/vit_mae.py [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+                    nn.init.constant_(m.bias, 0)
+                    nn.init.constant_(m.weight, 1.0)
+        else:
+            raise TypeError('pretrained must be a str or None')
+
+    def forward(self, x: Tensor) -> Tensor:
+        """Defines the computation performed at every call.
+
+        Args:
+            x (Tensor): The input data.
```
[the remark this answers is not in the document]
Comment (hukkai) [reply mmaction/models/backbones/vit_mae.py:354]: This is not valid, see pytorch: https://github.com/pytorch/pytorch/blob/6a2b12dd656ed8c347968bebb4c3552582454019/torch/nn/modules/activation.py#L57

File: mmaction/models/backbones/vit_mae.py [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+            state_dict = _load_checkpoint(self.pretrained)
+            if 'state_dict' in state_dict:
+                state_dict = state_dict['state_dict']
+            load_state_dict(self, state_dict, strict=False, logger=logger)
+        elif self.pretrained is None:
+            for m in self.modules():
+                if isinstance(m, nn.Linear):
+                    trunc_normal_init(m.weight, std=.02)
+                    if isinstance(m, nn.Linear) and m.bias is not None:
+                        nn.init.constant_(m.bias, 0)
```
[the remark this answers is not in the document]
Comment (hukkai) [reply mmaction/models/backbones/vit_mae.py:343]: Use torch.nn.init.trunc_normal_ instead.

File: mmaction/models/backbones/vit_mae.py [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+        self.qkv = nn.Linear(embed_dims, embed_dims * 3, bias=False)
+        self.attn_drop = nn.Dropout(attn_drop_rate)
+        self.proj = nn.Linear(embed_dims, embed_dims)
+        self.proj_drop = nn.Dropout(drop_rate)
+
+    def _init_qv_bias(self) -> None:
+        self.q_bias = nn.Parameter(torch.zeros(self.embed_dims))
+        self.v_bias = nn.Parameter(torch.zeros(self.embed_dims))
+
+    def forward(self, x: Tensor) -> Tensor:
```
[the remark this answers is not in the document]
Comment (hukkai) [reply mmaction/models/backbones/vit_mae.py:64]: This is not valid, see pytorch examples: https://github.com/pytorch/pytorch/blob/6a2b12dd656ed8c347968bebb4c3552582454019/torch/nn/modules/activation.py#L57

File: mmaction/models/backbones/vit_mae.py (hunk 5 above)
Comment (hukkai) [reply mmaction/models/backbones/vit_mae.py:247]: It should be dict or ConfigDict as suggested in mmengine

**Comment by hukkai** [comment]
Will open a new PR

File: mmaction/datasets/transforms/loading.py [hunk 10]
```diff
@@ -263,6 +263,84 @@ def __repr__(self):
         return repr_str
 
 
+@TRANSFORMS.register_module()
+class SampleFramesV2(SampleFrames):
```
Comment (ly015) [inline mmaction/datasets/transforms/loading.py:267]: I think the class name may need to be more informative and reflect the difference between this class and the base `SampleFrames`.

File: mmaction/datasets/transforms/loading.py [hunk 11]
```diff
@@ -174,9 +174,7 @@ def _get_train_clips(self, num_frames):
     def _get_test_clips(self, num_frames):
         """Get clip offsets in test mode.
 
-        Calculate the average interval for selected frames, and shift them
```
Comment (ly015) [inline mmaction/datasets/transforms/loading.py:177]: These descriptions are no longer needed?

**Review by ly015 — APPROVED** [review]

**Pull request merged**